### PR TITLE
Implement delete for HotDataSource

### PR DESCRIPTION
**Pull request by smatte**
## Description

It was not possible to delete items in HotDataSource cache

## Motivation and Context

In some case we need to invalidate/clear datasource to avoid getting previous request value. For instance when user logs out.

## How Has This Been Tested?

Unit test

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by CAptainDAniel** [comment]
There was the concept of clean(id), clean(list<id>) and cleanAll() in the BaseDatasource. Why was this not kept in the new HotDataSource ?

**Comment by martingagnon** [comment]
I don't really mind calling the new methods delete. However, delete(list) and deleteAll should be allowed.

File: datasources/src/commonTest/kotlin/com/mirego/trikot/datasources/BaseHotDataSourceTests.kt [hunk 1]
```diff
@@ -158,23 +174,29 @@ class BaseHotDataSourceTests {
...
 
     class CacheDataSource(private val readPromise: Promise<DataSourceTestData>) : BaseHotDataSource<TestDataSourceRequest, DataSourceTestData>() {
         var internalSaveCount = 0
+        var internalDeleteCount = 0
+        var deletedCacheableIds = mutableSetOf<Any>()
 
         override fun internalRead(request: TestDataSourceRequest): Promise<DataSourceTestData> {
-            return readPromise
+            return if (deletedCacheableIds.contains(request.cacheableId)) {
+                Promise.reject(Throwable())
```
Comment (marcantoinefortier) [inline datasources/src/commonTest/kotlin/com/mirego/trikot/datasources/BaseHotDataSourceTests.kt:186]: Maybe reject a Throwable with a clear explanation here? Let's say a custom `RequestCacheDeletedException : Exception("The cache was deleted for this request: $cacheableId")`

Comment (martingagnon) [reply datasources/src/commonTest/kotlin/com/mirego/trikot/datasources/BaseHotDataSourceTests.kt:186]: :this:

Comment (smatte) [reply datasources/src/commonTest/kotlin/com/mirego/trikot/datasources/BaseHotDataSourceTests.kt]: Done 👍

**Comment by smatte** [comment]
@CAptainDAniel @martingagnon 
> There was the concept of clean(id), clean(list) and cleanAll() in the BaseDatasource. Why was this not kept in the new HotDataSource ?

I added clean methods and tests.

**Comment by hugolefrancois** [comment]
> There was the concept of clean(id), clean(list) and cleanAll() in the BaseDatasource. Why was this not kept in the new HotDataSource ?

I didn't keep these methods because they could cause desynchronized data. Its will clean the data in memory in the data source, but it will only affect the future read calls. If someone is still listening on the publisher of a previous read, that publisher will never be updated again.

I think it's safer to not have these methods to avoid any bugs.

**Review by hugolefrancois — COMMENTED** [review]
Personally I don't like the concept of clean(id), clean(list) and cleanAll(), as I explained in my comment. But it's hop to you if you want to merge it, it could be useful in some situation if you really know what you are doing.

**Review by hugolefrancois — APPROVED** [review]

**Pull request merged**